articles/logging-with-winston/index.md

```diff
@@ -0,0 +1,575 @@
+
+When the server is running, accessing the following pages will create a log every time the link is invoked.
+
+- http://localhost:3000/ - the server will send a hello world message. We want Winston to capture that and record it in our log file.
+- http://localhost:3000/calc - we are trying to add variable `y` to variable `x`. In this case, variable `y` is not defined. This will generate an error, and we want Winston to capture such instances for us.
+- http://localhost:3000/hello - the server we have created has no such URL. We want Winston to inform us when a link that points to our IP address is accessed but can't be found; that is a `404` error.
+
+Go to the `server.log` file in the logs folder to view the log.
+Output:
+
+```js
+info: Nov-12-2020 10:07:59: 	Server started and running on http://localhost:3000
+info: Nov-12-2020 10:08:02: 	Server Sent A Hello World!
+error: Nov-12-2020 10:08:05: 	500 - Internal Server Error - y is not defined - /calc - GET - ::1
+error: Nov-12-2020 10:08:10: 	400 || Not Found - /hello - GET - ::1
+```
+
+Like we said earlier, Winston is well suited to configure different log destinations. In our small app, let's create another transport. This time, I want to save the logs to a database, [MongoDB](https://www.mongodb.com/try/download/community) to be concise. On the `app.js` file, replicate the following code block. Make sure you install [Winston MongoDB](https://www.npmjs.com/package/winston-mongodb), i.e., `npm install winston-mongodb`. If you are new to MongoDB, here's a [guide](/engineering-education/working-with-databases-part1/) that will help you get started.
+
+Replicate the following code blocks into the `logger.js` file.
+
+```js
+const { createLogger, format, transports } = require('winston');
+
+// Import mongodb
+require('winston-mongodb');
+
+module.exports = createLogger({
+transports:[
+
+// File transport
+    new transports.File({
+    filename: 'logs/server.log',
+    format:format.combine(
+        format.timestamp({format: 'MMM-DD-YYYY HH:mm:ss'}),
+        format.align(),
+        format.printf(info => `${info.level}: ${[info.timestamp]}: ${info.message}`),
+)}),
+
+// MongoDB transport
+    new transports.MongoDB({
+        level: 'error',
+        //mongo database connection link
+        db : 'mongodb://localhost:27017/logs',
+        options: {
+            useUnifiedTopology: true
+        },
+        // A collection to save json formatted logs
+        collection: 'server_logs',
+        format: format.combine(
+        format.timestamp(),
+        // Convert logs to a json format
+        format.json())
+    })]
+});
+```
+
+Run `node app.js` to start the server and access the following URLs to trigger server responses and requests.
+
+- http://localhost:3000/
+- http://localhost:3000/calc
+- http://localhost:3000/hello
+
+Logs will be recorded into the `server.log` file, and any `error` logs will be recorded in a MongoDB database.
```

Comment:
Would be nice if you provided instructions on how to see the log file. i.e. `mongo`, `use logs` and `db.server_logs.find()` in addition to linking to the MongoDB guide.